apisix/plugins/elasticsearch-logger.lua

```diff
@@ -0,0 +1,166 @@
+--
+
+local core            = require("apisix.core")
+local http            = require("resty.http")
+local log_util        = require("apisix.utils.log-util")
+local bp_manager_mod  = require("apisix.utils.batch-processor-manager")
+local plugin          = require("apisix.plugin")
+
+local ngx             = ngx
+local str_format      = core.string.format
+local str_byte        = string.byte
+
+local plugin_name = "elasticsearch-logger"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+
+
+local schema = {
+    type = "object",
+    properties = {
+        meta_format = {
+            type = "string",
+            default = "default",
+            enum = {"default", "origin"},
+        },
+        endpoint_addr = core.schema.uri_def,
+        field = {
+            type = "object",
+            properties = {
+                index = { type = "string"},
+                type = { type = "string"}
+            },
+            required = {"index"}
+        },
+        auth = {
+            type = "object",
+            properties = {
+                username = {
+                    type = "string",
+                },
+                password = {
+                    type = "string",
+                },
+            },
+            required = {"username", "password"},
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            default = 10
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        }
+    },
+    required = { "endpoint_addr", "field" },
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 413,
+    name = plugin_name,
+    schema = batch_processor_manager:wrap_schema(schema),
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function get_logger_entry(conf, ctx)
+    local entry
+    local metadata = plugin.plugin_metadata(plugin_name)
+    core.log.info("metadata: ", core.json.delay_encode(metadata))
+    if metadata and metadata.value.log_format
+        and core.table.nkeys(metadata.value.log_format) > 0
+    then
+        entry = log_util.get_custom_format_log(ctx, metadata.value.log_format)
+        core.log.info("custom log format entry: ", core.json.delay_encode(entry))
+    else
+        entry = log_util.get_full_log(ngx, conf)
+        core.log.info("full log entry: ", core.json.delay_encode(entry))
+    end
+
+    return core.json.encode({
+            create = {
+                _index = conf.field.index,
+                _type = conf.field.type
+            }
+        }) .. "\n" ..
+        core.json.encode(entry) .. "\n"
+end
+
+
+local function send_to_elasticsearch(conf, entries)
+    local httpc, err = http.new()
+    if not httpc then
+        return false, str_format("create http error: %s", err)
+    end
+
+    local uri = conf.endpoint_addr ..
+        (str_byte(conf.endpoint_addr, -1) == str_byte("/") and "_bulk" or "/_bulk")
```

Comment:
Look like we can require the endpoint_addr doesn't end with `/` in the schema. So that we don't need to check it  in per-request manner.

Reply:
That is a good idea!

Reply:
I define the `endpoint_addr`:
```
endpoint_addr = {
            type = "string",
            pattern = [=[^[^\/]+:\/\/([\da-zA-Z.-]+|\[[\da-fA-F:]+\])(:\d+)?/]=]
}
```
But if the `endpoint_addr` is set without "/", APISIX will return such error info:
```shell
{
  "error_msg": "failed to check the configuration of plugin elasticsearch-logger err: property \"endpoint_addr\" validation failed: failed to match pattern \"^[^\\\\/]+:\\\\/\\\\/([\\\\da-zA-Z.-]+|\\\\[[\\\\da-fA-F:]+\\\\])(:\\\\d+)?/\" with \"http://127.0.0.1:9200\""
}
```
It looks unfriendly to check the error, so I decide to use the method of checking whether there is a "/" in the code, or do you have a better solution?

Reply:
@ccxhwmy 
Sorry, I can't get your point. We can use a simple pattern that checks if the configuration doesn't end with `/`. Regularly, an `addr` should not end with `/`.

Reply:
> @ccxhwmy Sorry, I can't get your point. We can use a simple pattern that checks if the configuration doesn't end with `/`. Regularly, an `addr` should not end with `/`.

I have difficulty in doing it, would you give me some suggestion.

Reply:
@ccxhwmy 
What about:
```
addr = {
	type = "string",
	pattern = "[^/]$",
}
```

Reply:
> @ccxhwmy What about:
> 
> ```
> addr = {
> 	type = "string",
> 	pattern = "[^/]$",
> }
> ```

Thank you.